### ISSUE TITLE
linter: `prefer-optional-chain` false positive

### ISSUE THREAD

**Issue by lesha1201** [oxc-project/oxc#21077]
### What version of Oxlint are you using?

1.58.0

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "plugins": ["typescript"],
  "options": {
  	"typeAware": true
  },
  "rules": {
    "typescript/prefer-optional-chain": "deny"
  }
}
```

### What happened?

I've enocountered a false positive result of `prefer-optional-chain` on my project. Here is a simplified example of the issue:

```ts
const a: string | null = '';
const b: string = '';

// `prefer-optional-chain` rule complains about this code even though it doesn't make sense to use optional chain here
if (!a || b === a || b.indexOf('test')) {
  console.log('Hello, World');
}
```

```
  × typescript-eslint(prefer-optional-chain): Prefer using an optional chain expression instead, as it's more concise and easier to read.
    ╭─[<redacted>]
 11 │ 
 12 │ if (!a || b === a || b.indexOf('test')) {
    ·     ──────────────────────────────────
 13 │   console.log('Hello, World');
    ╰────
```

What's intersting is that if I change the order of variables in `===`, it stops complaining:

```ts
const a: string | null = '';
const b: string = '';

// Notice that now we have `a === b` instead of `b === a` and for some reason `prefer-optional-chain` stopped complaining on the code even though it didn't change logically
if (!a || a === b || b.indexOf('test')) {
  console.log('Hello, World');
}
```

typescript-eslint works fine for this case. Here is a playground that demonstrates it - [link](https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=MYewdgzgLgBAhgLhtATgSzAcxgHxmAVwBsiYBeGAckoG4BYAKFElgCMlUNsLr6HG0AMxgAKAIRxceSWVkxWU%2BQDoMAEwCmADwDygkZSjrolAJQmYAb0YwYzCCCLqlREJn0AJdSRAAaGAHUQFCJVUz4AX0ZGO1hgDih0LFx8YlIeWmjwaBhVeMTuKgz%2BBiFRMWBFVXI5Cpw8VRUwDR09AyMoU3MrBhs7BycXN0pPbz9A4NCTCKimLNh1PK5kwhJyQr4BYRF1GAAyXZgnNS1dfUNjM0trWyz%2B51cPLxcxoJCwxnCgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUh4AnonTCwAbSk0omdAENmkNWAC6UgL4hjQA&tokens=false).

**Comment by camc314** [comment]
0.20.0 got released yesterday - thanks for reporting this issue.

**Comment by lesha1201** [comment]
@camc314 Thank you very much for your work!
